docs/hugo/content/design/ADR-2022-12-Reconciliation-Extensions.md

```diff
@@ -0,0 +1,65 @@
+---
+title: '2022-12: Reconciliation Extensions'
+---
+
+## Context
+
+The current behaviour of Azure Service Operator (ASO) is to unilaterally issue a PUT to Azure each time reconciliation of a resource is triggered. 
+
+For some resources this doesn't work particularly well because the resources can be in a transient state that precludes the PUT from succeeding. The resulting error message pushes ASO into retrying the operation over and over again when it can't possibly succeed. Not only does this cause needless load on both the cluster and Azure, it can trigger rate limiting and other throttling mechanisms.
+
+Examples of this include:
+
+* An [AKS Managed Cluster](https://learn.microsoft.com/en-us/rest/api/aks/managed-clusters/get?tabs=HTTP) with `provisioningState` of `Updating`.
+* An [AKS Agent Pool](https://learn.microsoft.com/en-us/rest/api/aks/agent-pools/get?tabs=HTTP) with `provisioningState` of `Updating`.
+* A [PostgreSQL Flexible Server](https://learn.microsoft.com/en-us/rest/api/postgresql/flexibleserver/servers/get?tabs=HTTP) with a `state` of `Starting`, `Stopping` or `Updating`.
+
+This list not exhaustive.
+
+We need a way to allow the operator to handle these cases gracefully, by introducing an extension point where we can tweak the reconciliation flow and skip the PUT if the resource is in a state that we know will cause the PUT to fail.
+
+Complicating matters, we are also looking switch from using a unilateral PUT to a more discerning workflow where we do a GET first and only issue a PUT if the resource has changed (See [#2600](https://github.com/Azure/azure-service-operator/pull/2600) for the design). Ideally, our design for this extension point will be compatible with both the current PUT-only reconciliation flow and the new GET-PUT reconciliation flow. 
+
+In the most common scenario, this extension point will be used to skip reconciliation based on the current status of the resource. However, we also want to allow implementers to use this extension point to perform other actions, such as issuing a GET to check on the status of other Azure resources. 
+
+As with other extension points, we want to give implementers the option of acting either before or after the standard action.
+
+
+## Decision
+
+We'll define a new extension point called `ReconcileActionSelector` in a style similar to our already existing extension points.
+
+The extension point will receive the following parameters:
+
+* The current resource, with a status freshly updated from Azure.
+* A client allowing for ARM operations.
+* A logger to allow for tracing of what the extension did.
+* A context to allow cancellation of long running operations.
+* A function that will return the next action to take. 
```

Comment:
maybe, do you think an `Event` should also join the list?

Reply:
What would the event signal?

Reply:
A k8s events.. so when a user `k describe` a resource, they can see whether ASO issued a `GET` or a `PUT`.

Reply:
I don't think this is the place for that. 

This is the parameter list for the extension - and for it to work, it has to have the latest information, so at that point we'll always have just done a GET.